File: src/dtype.rs
use core::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DType {
    Bool,
    // integer data types
    Int8,
    Int16,
    Int32,
    Int64,
    // unsigned integer
    Uint8,
    // floating point datat types
    Float16,
    Float32,
    Float64,
    Bfloat16,
}

impl fmt::Display for DType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DType::Bool => write!(f, "bool"),
            DType::Int8 => write!(f, "int8"),
            DType::Int16 => write!(f, "int16"),
            DType::Int32 => write!(f, "int32"),
            DType::Int64 => write!(f, "int64"),
            DType::Uint8 => write!(f, "uint8"),
            DType::Float16 => write!(f, "float16"),
            DType::Float32 => write!(f, "float32"),
            DType::Float64 => write!(f, "float64"),
            DType::Bfloat16 => write!(f, "bfloat16"),
        }
    }
}

impl DType {
    pub fn get_bit_size(&self) -> isize {
        match self {
            DType::Bool => 1,
            DType::Int8 => 1,
            DType::Int16 => 2,
            DType::Int32 => 4,
            DType::Int64 => 8,
            DType::Uint8 => 1,
            DType::Float16 => 2,
            DType::Float32 => 4,
            DType::Float64 => 8,
            DType::Bfloat16 => 2,
        }
    }

    pub fn format_char(&self) -> char {
        // https://numpy.org/devdocs/reference/arrays.dtypes.html
        match self {
            DType::Bool => '?',
            DType::Int8 => 'b',
            DType::Int16 => 'h',
            DType::Int32 => 'i',
            DType::Int64 => 'q',
            DType::Uint8 => 'B',
            DType::Float16 => 'e',
            DType::Float32 => 'f',
            DType::Float64 => 'd',
            DType::Bfloat16 => 'v',
        }
    }

    pub fn from_str(string: &str) -> DType {
        match string {
            "?" => DType::Bool,
            "b" => DType::Int8,
            "h" => DType::Int16,
            "i" => DType::Int32,
            "q" => DType::Int64,
            "B" => DType::Uint8,
            "e" => DType::Float16,
            "f" => DType::Float32,
            "d" => DType::Float64,
            "v" => DType::Bfloat16,
            _ => panic!("DType {} is not supported.", string),
        }
    }
}

File: csrc/dtype.c
#include "dtype.h"
#include <assert.h>

size_t get_byte_size(DType dtype) {
  switch (dtype) {
    case Bool:
      return 1;
    case Int8:
      return 1;
    case Int16:
      return 2;
    case Int32:
      return 4;
    case Int64:
      return 8;
    case Uint8:
      return 1;
    case Float16:
      return 2;
    case Float32:
      return 4;
    case Float64:
      return 8;
    case Bfloat16:
      return 2;
  }
}

char format_char(DType dtype) {
  switch (dtype) {
    case Bool:
      return '?';
    case Int8:
      return 'b';
    case Int16:
      return 'h';
    case Int32:
      return 'i';
    case Int64:
      return 'q';
    case Uint8:
      return 'B';
    case Float16:
      return 'e';
    case Float32:
      return 'f';
    case Float64:
      return 'd';
    case Bfloat16:
      return 'v';
  }
}

DType dtype_from_str(char string) {
  switch (string) {
    case '?':
      return Bool;
    case 'b':
      return Int8;
    case 'h':
      return Int16;
    case 'i':
      return Int32;
    case 'q':
      return Int64;
    case 'B':
      return Uint8;
    case 'e':
      return Float16;
    case 'f':
      return Float32;
    case 'd':
      return Float64;
    case 'v':
      return Bfloat16;
    default:
      assert(0);
  }
}
